src/linear_algebra/free_module/ideal_quotient.lean

```diff
@@ -93,38 +97,97 @@ in (↑(e : (S ⧸ I) ≃ₗ[ℤ] _) : (S ⧸ I ≃+ _)).trans e'
 begin
+  haveI := fintype.of_finite ι,
   apply ((I.quotient_equiv_pi_span b _).restrict_scalars F).trans
     (direct_sum.linear_equiv_fun_on_fintype _ _ _).symm,
   exact linear_map.is_scalar_tower.compatible_smul
   -- why doesn't it automatically apply?
   -- even after `change linear_map.compatible_smul _ (Π i, R ⧸ span _) F R`
 end
 
-lemma ideal.finrank_quotient_eq_sum [nontrivial F]
-  [∀ i, module.free F (R ⧸ ideal.span ({I.smith_coeffs b hI i} : set R))]
-  [∀ i, module.finite F (R ⧸ ideal.span ({I.smith_coeffs b hI i} : set R))] :
+lemma finrank_quotient_eq_sum {ι} [fintype ι] (b : basis ι R S) [nontrivial F]
+  [∀ i, module.free F (R ⧸ span ({I.smith_coeffs b hI i} : set R))]
+  [∀ i, module.finite F (R ⧸ span ({I.smith_coeffs b hI i} : set R))] :
   finite_dimensional.finrank F (S ⧸ I)
-    = ∑ i, finite_dimensional.finrank F (R ⧸ ideal.span ({I.smith_coeffs b hI i} : set R)) :=
+    = ∑ i, finite_dimensional.finrank F (R ⧸ span ({I.smith_coeffs b hI i} : set R)) :=
 begin
-  rw [linear_equiv.finrank_eq $ ideal.quotient_equiv_direct_sum F b hI,
+  rw [linear_equiv.finrank_eq $ quotient_equiv_direct_sum F b hI,
       finite_dimensional.finrank_direct_sum]
   -- slow, and dot notation doesn't work
 end
+
+/-- For a nonzero element `f` in an algebra `S` over a principal ideal domain `R` that is finite and
+free as an `R`-module, the norm of `f` relative to `R` is associated to the product of the Smith
+coefficients of the ideal generated by `f`. -/
+lemma associated_norm_prod_smith {ι} [fintype ι] (b : basis ι R S) {f : S} (hf : f ≠ 0) :
+  associated (algebra.norm R f) (∏ i, smith_coeffs b _ (span_singleton_eq_bot.not.2 hf) i) :=
+begin
+  have hI := span_singleton_eq_bot.not.2 hf,
+  let b' := ring_basis b (span {f}) hI,
+  classical,
+  rw [← matrix.det_diagonal, ← linear_map.det_to_lin b'],
+  let e := (b'.equiv ((span {f}).self_basis b hI) $ equiv.refl _).trans
+    ((linear_equiv.coord S S f hf).restrict_scalars R),
+  refine (linear_map.associated_det_of_eq_comp e _ _ _).symm,
+  dsimp only [e, linear_equiv.trans_apply],
+  simp_rw [← linear_equiv.coe_to_linear_map, ← linear_map.comp_apply, ← linear_map.ext_iff],
+  refine b'.ext (λ i, _),
+  simp_rw [linear_map.comp_apply, linear_equiv.coe_to_linear_map, matrix.to_lin_apply,
+    basis.repr_self, finsupp.single_eq_pi_single, matrix.diagonal_mul_vec_single, pi.single_apply,
+    ite_smul, zero_smul, finset.sum_ite_eq', mul_one, if_pos (finset.mem_univ _), b'.equiv_apply],
+  change _ = f * _,
+  rw [mul_comm, ← smul_eq_mul, linear_equiv.restrict_scalars_apply, linear_equiv.coord_apply_smul,
+    ideal.self_basis_def],
+  refl
+end
+
+end comm_ring
+
+section field
+
+variables {F : Type*} [field F] [algebra F[X] S]
+
+instance (b : basis ι F[X] S) {I : ideal S} (hI : I ≠ ⊥) (i : ι) :
+  finite_dimensional F (F[X] ⧸ span ({I.smith_coeffs b hI i} : set F[X])) :=
+(adjoin_root.power_basis $ I.smith_coeffs_ne_zero b hI i).finite_dimensional
```

Comment:
This instance could be made more general but I suppose there would be no convenient way to do so yet. So it's probably fine.